[OpenReg][Feat][Docs] Enrich OpenReg device management implementation and add focused documentation

KarhouTam opened the pull request:
## Summary
This PR enriches OpenReg device management codes and adds focused documentation.

## Key Changes
- Introduced device management documentation in `device.md`.
- Updated `OpenRegFunctions.h` and `OpenRegFunctions.cpp` to use `DeviceIndex` and added error handling.
- Implemented `check_device_index` function for validating device indices.
- Enhanced Python bindings in `Module.cpp` for device management.
- Added tests for invalid device index handling in `test_device.py`.

KarhouTam commented on test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegFunctions.h:
```diff
@@ -9,10 +9,20 @@
 OPENREG_EXPORT DeviceIndex ExchangeDevice(DeviceIndex device);
 
+static inline void check_device_index(int64_t device) {
+  TORCH_CHECK((device >= 0 && device < c10::openreg::device_count()) || device == -1,
```
Sorry about that. I was thought that `-1` is the default and allowed. Fix already. Thanks!

fffrog commented on test/cpp_extensions/open_registration_extension/torch_openreg/torch_openreg/openreg/__init__.py:
```diff
@@ -42,7 +42,8 @@ def current_device():
 
 def set_device(device) -> None:
-    return torch_openreg._C._set_device(device)
+    if device >= 0:
```
Please don't let this failure happen silently.

KarhouTam commented on docs/source/accelerator/device.md:
```diff
@@ -0,0 +1,64 @@
+# Device Management
```
Will modify doc asap! Thanks for your reviewing.

KarhouTam committed:
[OpenReg] Enrich OpenReg device management documentation and implement device functions

- Introduced device management documentation in `device.md`.
- Updated `OpenRegFunctions.h` and `OpenRegFunctions.cpp` to use `DeviceIndex` and added error handling.
- Implemented `check_device_index` function for validating device indices.
- Enhanced Python bindings in `Module.cpp` for device management.
- Added tests for invalid device index handling in `test_device.py`.

KarhouTam commented:
Hi, @fffrog.  I think this PR should be merged after the hook one: #165980 . Cause there must have conflict in `index.md`, but this PR is enoughly fine to be merged (at least all CI passed).

KarhouTam commented:
@pytorchbot label "ciflow/trunk"

fffrog reviewed (approved):
LGTM, thank you a lot.

Just some tiny nits.

KarhouTam committed:
Add maybe_exchange_device function to OpenReg API for safe device switching

KarhouTam commented:
@pytorchbot label "ciflow/trunk"

KarhouTam committed:
Refactor device management documentation and update LITERALINCLUDE markers for SetDevice functions

KarhouTam committed:
Lint
